src/azul/service/manifest_service.py

```diff
@@ -776,6 +809,409 @@ def manifest_config(self) -> ManifestConfig:
+                                                        },
+                                                    },
+                                                },
+                                            ],
+                                        },
+                                    },
+                                },
+                                {
+                                    "name": "misc",
+                                    "type": {"type": "map", "values": "string"},
+                                },
+                            ],
+                        },
+                        *encoded_schema
+                    ]
+                },
+                {
+                    "name": "relations",
+                    "type": {
+                        "type": "array",
+                        "items": {
+                            "type": "record",
+                            "name": "Relation",
+                            "fields": [
+                                {"name": "dst_id", "type": "string"},
+                                {"name": "dst_name", "type": "string"},
+                            ],
+                        },
+                    },
+                    "default": [],
+                },
+            ],
+        }
+
+    nullable_to_pfb_types = {
+        null_bool: ['string', 'boolean'],
+        null_float: ['string', 'double'],  # Not present in current field_types
+        null_int: ['string', 'long'],
+        null_str: ['string'],
+        null_int_sum_sort: ['string', 'long']
+    }
+
+    @classmethod
+    def tables_schema(cls, field_types: FieldTypes) -> Iterable[JSON]:
+        """
+        Produce a sequence Avro schemas for each inner entity given a FieldTypes
+        mapping.
+
+        :param field_types: Dictionary structure mapping the ES doc structure to
+            types.
+        :return: A generator of schemas
+        """
+        for k, v in field_types.items():
+            yield {
+                "name": k,
+                "type": "record",
+                "fields": list(cls.tables_schema_recursive(v, singleton=k == 'files'))
+            }
+
+    @classmethod
+    def tables_schema_recursive(cls, field_types: FieldTypes, singleton: bool = False, depth=0) -> Iterable[JSON]:
+        # FIXME: Checking the type here is working around the unexpected
+        # attribute contents.total_estimated_cells
+        #  https://github.com/DataBiosphere/azul/issues/2298
+        if type(field_types) == dict:
+            for k, v in field_types.items():
+                if isinstance(v, dict):
+                    yield {
+                        "name": k,
+                        "type": {
+                            # This is always an array, even if singleton is passed in
+                            "type": "array",
+                            "items": {
+                                # FIXME: k should probably not be plural here
```

Comment:
As in ks? I don't understand.

Reply:
k is something like "files" or "specimens". In the schema we're describing the "name" (essentially the type) of one of entities in this array.

So intuitively you would expect an array called "files" to contain "records" called "file". Instead they're called "files", which is what this fixme is about.

Everything works fine the way it is, and this schema metadata is never displayed. But I put the FIXME in, because someone inspecting the PFB manually might find the naming unintuitive.